model_zoo/lbann.cpp

```diff
@@ -74,16 +69,19 @@ int main(int argc, char *argv[]) {
     stack_profiler::get()->activate(comm->get_rank_in_world());
 
     // Initalize a global I/O thread pool
-    std::shared_ptr<thread_pool> io_thread_pool = construct_io_thread_pool(comm);
+    std::shared_ptr<thread_pool> io_thread_pool = construct_io_thread_pool(comm.get());
```

Comment:
Maybe `auto` would be nicer here?

Reply:
`auto` *would* be nicer here, but it would be incorrect. `construct_io_thread_pool` returns `unique_ptr` (it's a factory function that doesn't store a reference to the created object, so `unique_ptr` is correct), and explicitly declaring the type casts to `shared_ptr` since the application code will be sharing that reference. One could make a stylistic argument that it should be
```cpp
auto io_thread_pool = std::shared_ptr<thread_pool>{construct_io_thread_pool(comm.get())};
```
They are the same and I have no strong feelings about the style difference. If there are opinions one way or the other, I'm happy to change it.

Reply:
No, if we can't do `auto`, then let's keep it as is.

Reply:
Ok, that's fine. For reference, the style shown in this comment is shown [here](https://herbsutter.com/2013/05/30/gotw-90-solution-factories), but there is no comment as to why that style might be preferred.